[QTL] Support multiple lookup maps within one namespace

sirpkt opened the pull request:
This PR is related with #2523 
- For URI or JDBC source with multiple columns, user can define all the needed (key column, value column) mappings within one namespace configuration.
- Impact on existing namespace implementations(like druid-Kafka-extraction-namespace) is minimized. Just change `implements ExtractionNamespaceFunctionFactory` to `extends ExtractionNamespaceFunctionFactory`.
- Now, NamespaceExtractor has one more parameter `mapName`, which indicates lookup map name in the given namespace, however, it works as before without that parameter for backward-compatibility.

fjy commented:
@sirpkt awesome!

b-slim commented:
@sirpkt i am not sure how this can work with the actual [LookupDimensionSpec](https://github.com/druid-io/druid/blob/4e119b7a24db6fd420c57489d8a472a278586968/processing/src/main/java/io/druid/query/dimension/LookupDimensionSpec.java#L66-L66) i am wondering how you will call it a query time.

sirpkt commented:
As I replied in the issue page,
user can specify lookup map name within the namespace like
```
{
  "type":"namespace",
  "namepace":"DB1",
  "mapName":"CtoD"
}
```
I added unit test code for explicit Json usage of NamepacedExtraction.

fjy commented:
@drcrallen @b-slim @cheddar can u guys review this and coordinate over development?

fjy commented:
@drcrallen @b-slim

drcrallen commented on extensions-core/namespace-lookup/src/main/java/io/druid/data/input/MultiMapsPopulator.java:
```diff
@@ -0,0 +1,90 @@
+              Map<K, V> map = maps.get(entry.getKey());
+              if (map == null)
+              {
+                map = new ConcurrentHashMap<K, V>();
```
This violates offheap caching

drcrallen commented on extensions-core/namespace-lookup/src/main/java/io/druid/server/namespace/cache/OffHeapNamespaceExtractionCacheManager.java:
```diff
@@ -157,7 +157,7 @@ public boolean delete(final String namespaceKey)
 
   @Override
-  public ConcurrentMap<String, String> getCacheMap(String namespaceOrCacheKey)
+  public ConcurrentMap<String, Map<String, String>> getCacheMap(String namespaceOrCacheKey)
```
I don't think this works at all as intended here.

drcrallen commented:
@sirpkt I think this one needs more discussion among the community to make sure it fits overall expectations. As such I'm proposing punting it out of 0.9.1.

0.9.1 is slated for a major overhaul of Lookups to essentially be the first (hopefully) production-ready version for lookups.

This is a (important) feature add for lookups, but is outside the scope of "required for MVP"

fjy commented:
@drcrallen @b-slim what's going on with this PR?

b-slim commented:
Same opinion as @drcrallen  the feature needs more discussion, plus major changes to be compatible with new lookups impls. In addition we have a pretty busy roadmap and i guess this feature is not a top priority IMHO. The author can always start working on make it working with new lookup module.

sirpkt commented:
I'll try to make this working with new lookup module.
I think it will take some time because lookup is changed considerably.

fjy commented:
@sirpkt @b-slim @drcrallen can we submit an issue or a proposal for the list of changes described in this PR and discuss changes there?

sirpkt commented:
I added `mapName` field in `LookupDimensionSpec` and `RegisteredLookupExtractionFn`
and updated docs to reflect the change of Globally Cached Lookups.
I also modified the description about Lookup Extraction Function
because LookupExtractor no longer supports `"namespace"` type.

jon-wei commented on extensions-core/lookups-cached-global/src/main/java/io/druid/server/lookup/namespace/cache/NamespaceExtractionCacheManager.java:
```diff
@@ -412,23 +431,79 @@ public void run()
+   * @return A Map<String, String>
    */
-  public abstract ConcurrentMap<String, String> getCacheMap(String namespaceOrCacheKey);
+  public Map<String, String> getCacheMap(String id, String mapName)
```
let's rename this to something like getCacheInnerMap() to differentiate the two functions, and note in the javadocs that this retrieves an inner map

jon-wei commented on extensions-core/lookups-cached-global/src/main/java/io/druid/query/lookup/namespace/NamespaceKey.java:
```diff
@@ -0,0 +1,66 @@
+/**
+ * NamespaceKey is used to distinguish the same mapNames of different namespace IDs
+ */
+public class NamespaceKey
```
For thought, would it be easier/better to use a MultiKey for the composite namespace:ID key?

sirpkt replied:
replaced by MultiKey

jon-wei commented on docs/content/development/extensions-core/lookups-cached-global.md:
```diff
@@ -185,22 +252,25 @@ The remapping values for each globally cached lookup can be specified by a json
 
 The `namespaceParseSpec` can be one of a number of values. Each of the examples below would rename foo to bar, baz to bat, and buck to truck. All parseSpec types assumes each input is delimited by a new line. See below for the types of parseSpec supported.
 
+The `keyValueMaps` specifies lookup maps under one lookup name. One lookup can have multiple maps that share the source of lookup. It is a list of maps called `keyValueMap` that has three entries, `mapName`, `keyName`, and `valueName`. `mapName` is the name of map inside the lookup and `keyName` is key column or field name and `valueName` is value column or field name.
```
Suggest adding a bit more documentation detail along the lines of:

- Key/Value column refer to columns within the lookup source; "columns" field refers to Druid columns whose values will be used as filtering criteria for retrieving the mapping row from the lookup source

sirpkt replied:
done

jon-wei commented on docs/content/development/extensions-core/lookups-cached-global.md:
```diff
@@ -185,22 +252,25 @@ The remapping values for each globally cached lookup can be specified by a json
-|`keyColumn`|The name of the column containing the key|no|The first column|
-|`valueColumn`|The name of the column containing the value|no|The second column|
+
+`columns` should contain all the columns specified in `kayValueMaps` of uri lookup spec.
```
typo: "kayValueMaps" -> "keyValueMaps"

jon-wei commented on docs/content/development/extensions-core/lookups-cached-global.md:
```diff
@@ -226,11 +294,10 @@ truck,something3,buck
 |`delimiter`|The delimiter in the file|no|tab (`\t`)|
 |`listDelimiter`|The list delimiter in the file|no| (`\u0001`)|
 
+`columns` should contain all the columns specified in `kayValueMaps` of uri lookup spec.
```
typo: "kayValueMaps"

sirpkt replied:
fixed as maps

jon-wei commented on docs/content/development/extensions-core/lookups-cached-global.md:
```diff
@@ -34,9 +34,22 @@ Globally cached lookups can be specified as part of the [cluster wide config for
+           "val2"
          ]
        },
+       "keyValueMaps":[
```
would it be simpler to call this list just "maps"?

sirpkt replied:
done

jon-wei commented on extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java:
```diff
@@ -180,7 +181,7 @@ public boolean start()
       final String topic = getKafkaTopic();
       LOG.debug("About to listen to topic [%s] with group.id [%s]", topic, factoryId);
-      final Map<String, String> map = cacheManager.getCacheMap(factoryId);
+      final Map<String, String> map = cacheManager.getCacheMap(factoryId, KeyValueMap.DEFAULT_MAPNAME);
```
Let's add a note in the docs about how KafkaLookupExtractor only uses the default mapname

sirpkt replied:
done

jon-wei commented on extensions-core/lookups-cached-global/src/main/java/io/druid/query/lookup/namespace/ExtractionNamespaceCacheFactory.java:
```diff
@@ -41,11 +44,32 @@
+    return getCachePopulator(id, extractionNamespace, lastVersion, cache);
+  }
+
+  public Callable<String> getCachePopulator(String id, T extractionNamespace, String lastVersion, Map<String, String> swap)
```
Can we add some javadocs explaining how this function differs from getMapCachePopulator()?

sirpkt replied:
done

jon-wei commented on extensions-core/lookups-cached-global/src/main/java/io/druid/server/lookup/namespace/cache/NamespaceExtractionCacheManager.java:
```diff
@@ -412,23 +431,79 @@ public void run()
 
   /**
-   * This method is expected to swap the cacheKey into the active namespace, and leave future requests for new cacheKey available. getCacheMap(cacheKey) should return empty data after this call.
+   * This method swap the cacheKey into the active namespace, and leave future requests for new cacheKey available. getCacheMap(cacheKey) should return empty data after this call.
```
spelling: "swap" -> "swaps", "leave" -> "leaves"

sirpkt replied:
done

jon-wei commented on extensions-core/lookups-cached-global/src/main/java/io/druid/server/lookup/namespace/cache/NamespaceExtractionCacheManager.java:
```diff
@@ -443,19 +518,35 @@ public void run()
+    }
   }
 
+  protected abstract boolean deleteInsideMap(final ConcurrentMap<String, Map<String, String>> map);
```
Let's add javadocs for these two methods

sirpkt replied:
done

jon-wei commented on extensions-core/lookups-cached-global/src/main/java/io/druid/server/lookup/namespace/cache/OnHeapNamespaceExtractionCacheManager.java:
```diff
@@ -55,51 +52,21 @@ public OnHeapNamespaceExtractionCacheManager(
-    finally {
-      lock.unlock();
-    }
+    // do nothing
```
Let's add a note on why the delete can be a no-op here (GC?)

sirpkt replied:
done

gianm commented on extensions-core/lookups-cached-global/src/main/java/io/druid/query/lookup/namespace/JDBCExtractionNamespace.java:
```diff
@@ -55,10 +55,8 @@ public JDBCExtractionNamespace(
-      @NotNull @JsonProperty(value = "valueColumn", required = true)
-      final String valueColumn,
+      @JsonProperty(value = "keyValueMaps", required = true)
+      KeyValueMap[] keyValueMaps,
```
Would prefer `List<KeyValueMap>` here, it's generally easier to work with.

sirpkt replied:
done

gianm commented on extensions-core/lookups-cached-global/src/main/java/io/druid/query/lookup/namespace/JDBCExtractionNamespace.java:
```diff
@@ -55,10 +55,8 @@ public JDBCExtractionNamespace(
       final String table,
-      @NotNull @JsonProperty(value = "keyColumn", required = true)
-      final String keyColumn,
-      @NotNull @JsonProperty(value = "valueColumn", required = true)
```
Perhaps we should have a "default map" name that the keyColumn/valueColumn go into if you don't specify a `maps` list. And then that one also gets used at query time if you don't specify a mapName.

gianm replied:
Ah, I see we already have this in `DEFAULT_MAPNAME`. Let's use that for this purpose.

sirpkt replied:
done

gianm commented on extensions-core/lookups-cached-global/src/main/java/io/druid/query/lookup/namespace/KeyValueMap.java:
```diff
@@ -0,0 +1,141 @@
+  @Override
+  public String toString()
+  {
+    return String.format("{map[%s] : key field = %s, value field = %s}", mapName, keyName, valueName);
```
Any reason not to use the auto generated IntelliJ style?

sirpkt replied:
changed to use auto generated one

gianm commented on extensions-core/lookups-cached-global/src/main/java/io/druid/server/lookup/namespace/JDBCExtractionNamespaceCacheFactory.java:
```diff
@@ -76,53 +81,84 @@ public String call()
+  private String makeQueryString(List<String> requiredFields)
+  {
+    String query = "SELECT ";
+    query += StringUtils.join(requiredFields, ',');
```
These should be escaped; field names can have funny characters in them.

sirpkt replied:
done

gianm commented on extensions-core/lookups-cached-global/src/main/java/io/druid/query/lookup/namespace/KeyValueMap.java:
```diff
@@ -0,0 +1,141 @@
+  private final String keyName;
+  private final String valueName;
+
+  public static String DEFAULT_MAPNAME = "default";
```
`__default` is more consistent with defaults in other Druid areas.

gianm commented on docs/content/development/extensions-core/lookups-cached-global.md:
```diff
@@ -34,9 +34,22 @@ Globally cached lookups can be specified as part of the [cluster wide config for
+         {
+           "mapName":"default",
+           "keyColumn":"key",
+           "valueColumn":"val1"
```
When reading through the URIExtractionNamespace changes I now wonder if having the `"maps"` with their keyColumn / valueColumn out here is causing the dizziness and weirdness with simpleJson… because it has no keyColumn!

I wonder if it makes more sense to move `"maps"` into the namespaceParseSpec. That way the parser is in charge of what the map names and k/vs it returns are, and that should remove some of the weirdness in URIExtractionNamespace.

jon-wei replied:
that seems like a reasonable change, it would express more directly that simpleJson parser doesn't use the "maps" field unlike the other parser types

I suppose the logic for map building from a set of KeyValueMaps in the URIExtractionNamespace's delegate parser could be moved to something shared by the CSV/TSV/customJson "FlatDataParsers" in URIExtractionNamespace

gianm commented on extensions-core/lookups-cached-global/src/main/java/io/druid/query/lookup/namespace/URIExtractionNamespace.java:
```diff
@@ -185,45 +199,64 @@ public int hashCode()
-      return ImmutableMap.<String, String>of(k, val);
+      ImmutableMap.Builder<MultiKey, Map<String, String>> builder = new ImmutableMap.Builder<>();
+
+      // somewhat dizzy to support simple json case
```
See comment above… I wonder if it'd be less dizzy to move `"maps"` into the namespaceParseSpec.

gianm commented on extensions-core/lookups-cached-global/src/main/java/io/druid/server/lookup/namespace/JDBCExtractionNamespaceCacheFactory.java:
```diff
@@ -76,53 +81,84 @@ public String call()
+  private String makeQueryString(List<String> requiredFields)
+  {
+    String query = "SELECT \"";
+    query += StringUtils.join(requiredFields, "\", \"");
```
I don't think this is enough escaping. There could be backslashes and quotes and stuff in the field names. Maybe. Does JDBC/JDBI have a utility function to help with escaping?

EIther that, or let's check the requiredFields against a whitelist of characters.

gianm commented on docs/content/development/extensions-core/lookups-cached-global.md:
```diff
@@ -34,9 +34,22 @@ Globally cached lookups can be specified as part of the [cluster wide config for
        },
+       "maps":[
+         {
+           "mapName":"default",
```
Please also document this as the default map name.

gianm commented:
@drcrallen @b-slim any thoughts on the general idea & API here?

Broadly: looks like the changes are all centered around having more than one lookup map per thing-we-load. So we may load a single json file that has many logical lookups in it. IMO the nice thing about doing it this way is we only have to poll and parse the file one time. It's also easier to configure loading multiple lookups from one file. I'm on board with the general idea and attempting to work out whether the API needs adjustments or not.

b-slim commented on processing/src/main/java/io/druid/query/dimension/LookupDimensionSpec.java:
```diff
@@ -70,6 +75,7 @@ public LookupDimensionSpec(
       @JsonProperty("retainMissingValue") boolean retainMissingValue,
       @JsonProperty("replaceMissingValueWith") String replaceMissingValueWith,
       @JsonProperty("name") String name,
+      @JsonProperty("mapName") String mapName,
```
why is this changing the API ? not all the lookups will have a map name ?

b-slim commented:
I like the idea on minimizing the amount of fetch that a lookup had to make but the current API change make it backward incompatible plus it is unclear what `mapName` really mean. I would highly recommend to work on that by for instance use a name spacing convention like `name.subname` where the name will be used to match on the registered lookup then subname as equivalent to `mapName`.

sirpkt commented:
Sorry for late response.
I updated the code based on the review comments.

@b-slim I don't understand your point about backward compatibility because `mapName` is optional argument and users always make LookupDimensionSpec with json  so that they just omit
 `mapName` in their spec when their lookups do not have multiple maps.
And I still think having separate `mapName` is better than combining `name` and `mapName` because user may use combining delimiter (ex. `.`) in `name` or `mapName`.
However, it makes sense that `mapName` is unclear so I changed it to `innerMapName`. Welcome any suggestion.

@gianm For escaping column and table names at SQL query creation, I use escape and quote methods of SQLTemplate in Querydsl. As I'm not familiar with SQL querying in Java, I'm not sure that this make sense.

Other updates:
- KeyValueMap is moved to namespaceParseSpec from URIExtractionNamespace.
- Dependency on commons-collections is removed by using Pair
- FlatDataParser is refactored
- NamespaceLookupIntrospectHandler is modified as suggested by @gianm

fjy commented:
@b-slim @gianm can we finish this up?

gianm commented:
Moving to 0.10.1 as review is not complete. @sirpkt please let us know if you're still interested and we will endeavor to take another look.
